pages/bridge.tsx

```diff
@@ -134,12 +224,54 @@ const Bridge = () => {
                 {mounted ? formatUnits(source.fee, 18) : 0} ETH
               </div>
               <div className="text-center">
-                <button
-                  type="button"
-                  className="mt-5 rounded-md bg-indigo-600 px-3.5 py-2.5 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
-                >
-                  Bridge to {target.chain.name}
-                </button>
+                {bridgeTarget === BridgeTarget.L2 && needsAllowanceL1 ? (
+                  <button
+                    type="button"
+                    className="mt-5 rounded-md bg-indigo-600 px-3.5 py-2.5 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 disabled:bg-gray-300 disabled:cursor-not-allowed"
+                    onClick={handleAllowance}
+                    disabled={!approveL1}
+                  >
+                    Set allowance for {target.token?.symbol} on{" "}
+                    {target.chain.name}
+                    {approveL1IsLoading && <Spinner />}
+                  </button>
+                ) : (
+                  <button
+                    type="button"
+                    className="mt-5 mx-auto flex flex-row items-center rounded-md bg-indigo-600 px-3.5 py-2.5 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 disabled:bg-gray-300 disabled:cursor-not-allowed"
+                    onClick={handleBridge}
+                    disabled={
+                      bridgeTarget === BridgeTarget.L2
+                        ? !bridgeToL2
+                        : !bridgeToL1
+                    }
+                  >
+                    <span>Bridge to {target.chain.name}</span>
+                    {((bridgeTarget === BridgeTarget.L2 &&
+                      bridgeToL2IsLoading) ||
+                      (bridgeTarget === BridgeTarget.L1 &&
+                        bridgeToL1IsLoading)) && (
+                      <div className="ml-2">
+                        <Spinner />
+                      </div>
+                    )}
+                  </button>
+                )}
+              </div>
+              <div className="mt-5">
+                {bridgeTarget === BridgeTarget.L2
```

Comment:
We can eliminate one class of issues if we prompt the user to switch their chain. Either when they switch are try to bridge while on the wrong chain. I have an example of switching chains in the delegate page pr.